f5_openstack_agent/utils/cert_parser.py

```diff
@@ -0,0 +1,208 @@
+    state = stSpam
+    if type(data) == six.binary_type:
+        data = data.decode('utf-8')
+    for certLine in data.replace('\r', '').split('\n'):
+        if not certLine:
+            continue
+        certLine = certLine.strip()
+        if state == stSpam:
+            if certLine in startMarkers:
+                certLines = []
+                idx = startMarkers[certLine]
+                state = stHam
+                continue
+        if state == stHam:
+            if certLine in stopMarkers and stopMarkers[certLine] == idx:
+                state = stDump
+            else:
+                certLines.append(certLine)
+        if state == stDump:
+            yield b''.join([base64.b64decode(x) for x in certLines])
+            state = stSpam
+
+
+def _get_certs_from_pkcs7_substrate(substrate):
+    """Extracts DER-encoded X509 certificates from a PKCS7 ASN1 DER substrate
+
+    :param substrate: The substrate to be processed
+    :returns: A list of DER-encoded X509 certificates
+    """
+    try:
+        contentInfo, _ = der_decoder.decode(substrate,
+                                            asn1Spec=rfc2315.ContentInfo())
+        contentType = contentInfo.getComponentByName('contentType')
+    except Exception:
+        LOG.exception('Unreadable Certificate.')
+        raise f5_ex.UnreadableCert
+    if contentType != rfc2315.signedData:
+        LOG.exception('Unreadable Certificate.')
+        raise f5_ex.UnreadableCert
+
+    try:
+        content, _ = der_decoder.decode(
+            contentInfo.getComponentByName('content'),
+            asn1Spec=rfc2315.SignedData())
+    except Exception:
+        LOG.exception('Unreadable Certificate.')
+        raise f5_ex.UnreadableCert
+
+    for cert in content.getComponentByName('certificates'):
+        yield der_encoder.encode(cert)
+
+
+def _get_x509_from_pem_bytes(certificate_pem):
+    """Parse X509 data from a PEM encoded certificate
+
+    :param certificate_pem: Certificate in PEM format
+    :returns: crypto high-level x509 data from the PEM string
+    """
+    if type(certificate_pem) == six.text_type:
+        certificate_pem = certificate_pem.encode('utf-8')
+    try:
+        x509cert = x509.load_pem_x509_certificate(certificate_pem,
+                                                  backends.default_backend())
+    except Exception:
+        LOG.exception('Unreadable Certificate.')
+        raise f5_ex.UnreadableCert
+    return x509cert
+
+
+def _get_x509_from_der_bytes(certificate_der):
+    """Parse X509 data from a DER encoded certificate
+
+    :param certificate_der: Certificate in DER format
+    :returns: crypto high-level x509 data from the DER-encoded certificate
+    """
+    try:
+        x509cert = x509.load_der_x509_certificate(certificate_der,
+                                                  backends.default_backend())
+    except Exception:
+        LOG.exception('Unreadable Certificate.')
+        raise f5_ex.UnreadableCert
+    return x509cert
```

Comment:
Do we want to add some basic test coverage for this `cert_parser`?

Reply:
I am not sure that we need to add all test cases in the future code for this repo?

When I run the tox unit test, I found some of the previous tests (test_l2_service.py, test_network_helper_delete_fdb_entries) ran wrong, and I am sure it's not caused by my code changes. 

I will fix it as a bug if no one to fix it.